src/features/broker-protection/actions/extract.js

```diff
@@ -0,0 +1,300 @@
+import { getElement, getElementMatches, getElements } from '../utils.js' // Assuming you have imported the address comparison function
+import { SuccessResponse } from '../types.js'
+import { isSameAge } from '../comparisons/is-same-age.js'
+import { isSameName } from '../comparisons/is-same-name.js'
+import { matchesFullAddress } from '../comparisons/matches-full-address.js'
+import { matchAddressFromAddressListCityState } from '../comparisons/address.js'
+
+/**
+ * Adding these types here so that we can switch to generated ones later
+ * @typedef {Record<string, any>} Action
+ */
+
+/**
+ * @typedef {Object} ExtractProfileProperty
+ * For example: {
+ *   "selector": ".//div[@class='col-sm-24 col-md-8 relatives']//li"
+ * }
+ * @property {string} selector - xpath or css selector
+ * @property {boolean} [findElements] - whether to get all occurrences of the selector
+ * @property {string} [afterText] - get all text after this string
+ * @property {string} [beforeText] - get all text before this string
+ * @property {string} [separator] - split the text on this string
+ */
+
+/**
+ * @param {Action} action
+ * @param {Record<string, any>} userData
+ * @return {import('../types.js').ActionResponse}
+ */
+export function extractProfiles (action, userData) {
+    const profilesElementList =
+      Array.from(document.querySelectorAll(action.selector)) ?? []
+
+    const matchedProfiles = profilesElementList
+        // first, convert each profile element list into a profile
+        .map((element) => createProfile(element, action.profile))
+        // only include profiles that match the user data
+        .filter((scrapedData) => scrapedDataMatchesUserData(userData, scrapedData))
+        // aggregate some fields
+        .map((scrapedData) => aggregateFields(scrapedData))
+
+    return new SuccessResponse({ actionID: action.id, actionType: action.actionType, response: matchedProfiles })
+}
+
+/**
+ * Produces structures like this:
+ *
+ * {
+ *   "name": "John V Smith",
+ *   "alternativeNamesList": [
+ *     "John Inc Smith",
+ *     "John Vsmith",
+ *     "John Smithl"
+ *   ],
+ *   "age": "97",
+ *   "addressCityStateList": [
+ *     {
+ *       "city": "Orlando",
+ *       "state": "FL"
+ *     }
+ *   ],
+ *   "profileUrl": "https://example.com/1234"
+ * }
+ *
+ * @param {HTMLElement} profileElement
+ * @param {Record<string, ExtractProfileProperty>} extractData
+ * @return {Record<string, any>}
+ */
+function createProfile (profileElement, extractData) {
+    const output = {}
+    for (const [key, value] of Object.entries(extractData)) {
+        if (!value?.selector) {
+            output[key] = null
+        } else {
+            const evaluatedValue = value?.findElements
+                ? findFromElements(profileElement, key, value)
+                : findFromElement(profileElement, key, value)
+
+            const extractedValue = extractValue(key, value, evaluatedValue)
```

Comment:
hm, `evalueatedValue` may be an array (if it's produced by `findFromElements()`), but `extractValue` seems to expect a single string (at least in some extractors such as `name`). Am I missing something?

Reply:
No, I think this just needs be a bit more explicit, let me re-work it

Reply:
I altered the types + added a comment to make it explicit. 

IMO, this needs re-visiting at a later date, there's quite a mix of types on inputs/output and it's making it difficult to review.